File: js/src/mcp/resources/specsResource.ts
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { McpResource } from '../types.js';

const SPEC_FILES = [
  { id: 'journeys', filename: 'journeys.yaml', mimeType: 'application/yaml' },
  { id: 'selectors', filename: 'selectors.md', mimeType: 'text/markdown' },
  { id: 'testing-philosophy', filename: 'testing-philosophy.md', mimeType: 'text/markdown' }
];

export const loadSpecResources = (): McpResource[] => {
  const repoRoot = findRepoRoot(path.dirname(fileURLToPath(import.meta.url)));
  const specDir = path.join(repoRoot, 'common', 'specs');

  return SPEC_FILES.map(({ id, filename, mimeType }) => {
    const filePath = path.join(specDir, filename);
    const content = fs.readFileSync(filePath, 'utf-8');
    return {
      id,
      name: filename,
      path: filePath,
      mimeType,
      content
    } satisfies McpResource;
  });
};

const findRepoRoot = (startDir: string): string => {
  let currentDir = startDir;

  while (true) {
    if (fs.existsSync(path.join(currentDir, 'common', 'specs'))) {
      return currentDir;
    }

    const parentDir = path.dirname(currentDir);
    if (parentDir === currentDir) {
      throw new Error(`Unable to locate common/specs from ${startDir}`);
    }
    currentDir = parentDir;
  }
};

File: py/src/playwright_adventures/mcp_server/resources/specs_resource.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import List


@dataclass
class SpecResource:
    id: str
    name: str
    mime_type: str
    path: str
    content: str


SPEC_FILES = [
    ("journeys", "journeys.yaml", "application/yaml"),
    ("selectors", "selectors.md", "text/markdown"),
    ("testing-philosophy", "testing-philosophy.md", "text/markdown"),
]


def _repo_root() -> Path:
    return Path(__file__).resolve().parents[5]


def load_spec_resources() -> List[SpecResource]:
    spec_dir = _repo_root() / "common" / "specs"
    resources: List[SpecResource] = []
    for resource_id, filename, mime_type in SPEC_FILES:
        file_path = spec_dir / filename
        content = file_path.read_text(encoding="utf-8")
        resources.append(
            SpecResource(
                id=resource_id,
                name=filename,
                mime_type=mime_type,
                path=str(file_path),
                content=content,
            )
        )
    return resources
